File: smallest_number_in_infinite_set.js
var SmallestInfiniteSet = function() {
    this.heap = [];
    this.min_num = 1;
};

/**
 * @return {number}
 */
SmallestInfiniteSet.prototype.popSmallest = function() {
    if (this.heap.length > 0) {
        return this.heap.shift();
    }
    this.min_num += 1;
    return this.min_num - 1;    
};

/** 
 * @param {number} num
 * @return {void}
 */
SmallestInfiniteSet.prototype.addBack = function(num) {
    if (this.min_num > num && this.heap.indexOf(num) === -1) {
        this.heap.push(num);
        this.heap.sort((a, b) => a - b);
    }
};

/** 
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * var obj = new SmallestInfiniteSet()
 * var param_1 = obj.popSmallest()
 * obj.addBack(num)
 */

File: smallest_number_in_infinite_set.java
class SmallestInfiniteSet {
    private PriorityQueue<Integer> heap;
    private int min_num;

    public SmallestInfiniteSet() {
        heap = new PriorityQueue<>();
        min_num = 1;        
    }
    
    public int popSmallest() {
        if (!heap.isEmpty()) {
            return heap.poll();
        }
        min_num += 1;
        return min_num - 1;        
    }
    
    public void addBack(int num) {
        if (min_num > num && !heap.contains(num)) {
            heap.offer(num);
        }        
    }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet obj = new SmallestInfiniteSet();
 * int param_1 = obj.popSmallest();
 * obj.addBack(num);
 */
